content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+  
+  ```shell
+  # kubectl auth can-i create deployments --namespace dev
+  ```
+  
+  When you run the command above, Kubernetes creates a `SelfSubjectAccessReview` resource in the
+  Kubernetes API. This allows Kubernetes to fetch a list of possible permissions for your user.
+  Kubernetes then generates a personalized response to your request in real-time. This logic is
+  different from a scenario where this resource is simply stored in etcd.
+  {{% /alert %}}
+
+- Similarly, in KubeVirt's [CDI (Containerized Data Importer)](https://github.com/kubevirt/containerized-data-importer)
+  extension, which allows file uploads into a PVC from a local machine using the `virtctl` tool,
+  a special token is required before the upload process begins.
+  This token is generated by creating an `UploadTokenRequest` resource in the Kubernetes API. Kubernetes
+  routes all `UploadTokenRequest` resources creation requests to the CDI api-server,
+  which generates and returns the token in response.
+
+### Full control over conversion, validation, and output formatting
+
+- Your own api-server has all the capabilities of the vanilla Kubernetes api-server. The resources you create
+  in your api-server can be validated immediately on the server side without additional webhooks
+  (with CRDs, you'd have to write webhooks for extended validation).
+  
+  You can implement transparent conversion logic between versions like `v1alpha1`, `v1beta1`, and `v1`
+  and serve multiple API versions simultaneously.
+  
+  
+  {{% alert title="Note" color="info" %}}
+  Kubernetes allows you to serve multiple API versions for each resource type, traditionally
+  `v1alpha1`, `v1beta1` and `v1`. Only one version can be specified as the storage version.
+  All requests to other versions must be automatically converted to the version specified as storage version.
+  With CRDs, this mechanism is implemented using conversion webhooks. Whereas in extension api-server,
+  this logic can be implemented directly.
+  {{% /alert %}}
+
+- The api-server lets you format table output however you like and doesn't force you to follow
+  the `additionalPrinterColumns` logic in CRDs. Instead, you can write your own formatter that
+  formats the table output and custom fields in it. For example, when using `additionalPrinterColumns`,
+  you can display field values only following the JSONPath logic. In your own api-server, you can generate
+  and insert values on the fly, formatting the table output as you wish.
+
+### Dynamic resource registration
+
+- The resources served by an extension api-server don't need to be pre-registered as CRDs.
+Once your api-server is registered as an `APIService`, Kubernetes starts polling it to find out what
+resource types it serves. After receiving a response, it automatically registers all available
+types in its API. Consequently, you can implement logic that automatically registers the resource
+types you need in Kubernetes.
```

Comment:
This doesn't quite make sense. You can serve any discovery document for an API group, but if you try to serve a different API group, I don't think that's going to work.

Also, you do need to serve an API discovery document - or at least, that's what we want readers to see as recommended practice.